### PR TITLE
check the error message to only handle the expected exception

### PR DESCRIPTION
**Pull request by TingDaoK**
fix Misleading error message when mqtt on_message_received callback raises a TypeError by checking the error message.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: awscrt/mqtt.py [hunk 1]
```diff
@@ -499,11 +499,14 @@ def subscribe(self, topic, qos, callback=None):
             def callback_wrapper(topic, payload, dup, qos, retain):
                 try:
                     callback(topic=topic, payload=payload, dup=dup, qos=QoS(qos), retain=retain)
-                except TypeError:
-                    # This callback used to have fewer args.
-                    # Try again, passing only those those args, to cover case where
-                    # user function failed to take forward-compatibility **kwargs.
-                    callback(topic=topic, payload=payload)
+                except TypeError as e:
+                    if "got an unexpected keyword argument" in str(e) and callback.__name__ in str(e):
```
Comment (ilevyor) [inline awscrt/mqtt.py:503]: Is there a more robust way to handle this? What if someone changes the string value of the error? I don't think they would expect that to affect people's code execution beyond logging.

File: awscrt/mqtt.py [hunk 2]
```diff
@@ -499,11 +499,14 @@ def subscribe(self, topic, qos, callback=None):
...
+                        # This callback used to have fewer args.
+                        # Try again, passing only those those args, to cover case where
```
Comment (ilevyor) [inline awscrt/mqtt.py:505]:
```suggestion
                        # Try again, passing only those args, to cover case where
```

File: awscrt/mqtt.py [hunk 3]
```diff
@@ -557,11 +560,15 @@ def on_message(self, callback):
...
```
Comment (ilevyor) [inline awscrt/mqtt.py:566]:
```suggestion
                        # Try again, passing only those args, to cover case where
```

File: awscrt/mqtt.py [hunk 4]
```diff
@@ -322,6 +322,19 @@ def __init__(self,
...
+    def _check_try_backward_message_callback(self, callback):
+        # The callback used to have fewer args. Passing only those args, if it
+        # only has two args and no forward-compatibility to cover case where
+        # user function failed to take forward-compatibility **kwargs.
+
+        try_backward_callback = False
+        callback_sig = signature(callback)
+        try:
+            callback_sig.bind(topic='topic', payload='payload', dup=True, qos=QoS(1), retain=True)
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (graebm) [reply awscrt/mqtt.py:334]: optional: before this PR we didn't know it was possible to check signatures at all

now that we do, let's proactively check the callback will work with the old signature as well. maybe like:
```
try:
    # try new signature
    callback_sig.bind(topic='topic', payload='payload', dup=True, qos=QoS(1), retain=True)
    return False
except TypeError:
    # try old signature
    callback_sig.bind(topic='topic', payload='payload')
    return True
```

**Review by graebm — APPROVED** [review]
fix & ship

**Pull request merged**